### PR TITLE
chore: replace govendor with go mod and update docs

### PR DESCRIPTION
**Pull request by yeya24**
### Ⅰ. Describe what this PR did
Replace govendor with go mod since we have already updated to go 1.12.6.

### Ⅱ. Does this pull request fix one issue?

fix second part of #634

### CODE REVIEW

**Comment by allencloud** [comment]
It is a big change to the project.
I am wondering if there is any possibility to check the go mod file's correctness via Makefile target?
If we can do that, we could validate the dependency's correctness.
Just my little opinion. Wish your feedback. @yeya24

**Comment by yeya24** [comment]
> It is a big change to the project.
> I am wondering if there is any possibility to check the go mod file's correctness via Makefile target?
> If we can do that, we could validate the dependency's correctness.
> Just my little opinion. Wish your feedback. @yeya24

@allencloud 
Yes this validation is important. And I see this Makefile in [Prometheus](https://github.com/prometheus/promu/blob/3e43dfdc9726cee3f7c21476d9add8f13bebb645/Makefile.common#L188) is like doing what we want. Is this correct?
I change title to WIP and I will try to work on this.

**Comment by inoc603** [comment]
Since we're hopping on the go mod train, maybe we can consider not checking the `vendor/` directory into git repo. `go.sum` and `go.mod` can pin dependency versions. Removing `vendor/` will prevent huge PRs when changing dependencies. Also we don't have to validate whether `vendor/` and `go.sum` matches.

Of course not checking in `vendor/` means we would spend more time downloading dependencies during CI. This can be solved by either:
- Make an image with dependencies, which is automatically updated when `go.sum` and `go.mod` changes, and do compiling in it.
- Or use a proxy that supports the [module proxy protocol](https://tip.golang.org/cmd/go/#hdr-Module_proxy_protocol), some options can be found [here](https://github.com/golang/go/wiki/Modules#are-there-always-on-module-repositories-and-enterprise-proxies)

**Comment by yeya24** [comment]
@allencloud  @lowzj  WDYT?

**Comment by Starnop** [comment]
@lowzj @allencloud @inoc603 Now that we have updated the go version to 1.12.6, so let's make this PR as a break change for release 0.4.1. WDYT?

**Comment by Starnop** [comment]
> Since we're hopping on the go mod train, maybe we can consider not checking the `vendor/` directory into git repo. `go.sum` and `go.mod` can pin dependency versions. Removing `vendor/` will prevent huge PRs when changing dependencies. Also we don't have to validate whether `vendor/` and `go.sum` matches.
> 
> Of course not checking in `vendor/` means we would spend more time downloading dependencies during CI. This can be solved by either:
> 
> * Make an image with dependencies, which is automatically updated when `go.sum` and `go.mod` changes, and do compiling in it.
> * Or use a proxy that supports the [module proxy protocol](https://tip.golang.org/cmd/go/#hdr-Module_proxy_protocol), some options can be found [here](https://github.com/golang/go/wiki/Modules#are-there-always-on-module-repositories-and-enterprise-proxies)

I agree with that. Even though `go mod` now supports the vendor, however, it will be abandoned in the future. And as @inoc603 said, we can use docker image to provide repeatable environments.

**Comment by yeya24** [comment]
Currently I met some problems in fixing ci.
- gometalinter cannot support go mod without vendor. See issue https://github.com/alecthomas/gometalinter/issues/562. Currently I have to run  go mod vendor first then run gometalinter. Maybe we can update to golang-ci in the future since gometalinter has deprecated.
- It will always fail in integration-test. I am still finding the cause.

File: Makefile [hunk 1]
```diff
@@ -94,13 +100,33 @@ unit-test: build-dirs
 	./hack/unit-test.sh
 .PHONY: unit-test
 
+integration-test:
+	@go build ./test
```
Comment (Starnop) [inline Makefile]: s/ `@go build ./test`/ `@go test ./test` ? Maybe it's the cause of the failure of the integration test.

File: Makefile [hunk 2]
```diff
@@ -94,13 +100,33 @@ unit-test: build-dirs
...
+go-mod-tidy:
+	@echo "Begin to tidy up go.mod and go.sum"
+	@go mod tidy
+.PHONY: go-mod-tidy
+
+# we need this because currently gometalinter can't work in go mod environment.
+go-mod-vendor:
+	@echo "Begin to vendor go mod dependency"
+	@go mod vendor
+.PHONY: go-mod-tidy
```
Comment (Starnop) [inline Makefile]: go-mod-vendor？

File: hack/env.sh [hunk 3]
```diff
@@ -24,3 +24,6 @@ GOOS=$(go env GOOS)
 GOARCH=$(go env GOARCH)
 export GOOS
 export GOARCH
+
+export GO111MODULE=on
+export GOPROXY=https://goproxy.io
```
Comment (Starnop) [inline hack/env.sh]: Add a newline at end of file.

File: Makefile (hunk 2 above)
Comment (yeya24) [reply Makefile]: Yes, see this issue alecthomas/gometalinter#562. Currently gometalinter cannot work in go mod env without vendor. So I have to add this during ci when checking gometalinter.

File: hack/env.sh (hunk 3 above)
Comment (yeya24) [reply hack/env.sh]: Done.

File: Makefile (hunk 2 above)
Comment (yeya24) [reply Makefile]: Done.

File: Makefile (hunk 1 above)
Comment (yeya24) [reply Makefile]: Thanks for this help!

**Comment by yeya24** [comment]
@Starnop @lowzj  Can you review this again? Pass the ci now.

**Comment by Starnop** [comment]
LGTM. And @lowzj  PTAL again.

**Comment by Starnop** [comment]
LGTM.

**Comment by allencloud** [comment]
Great work. We are moving to `go mod`. 👍 @yeya24 @Starnop

**Pull request merged**